src/wordlift/mappings/class-mappings-validator.php

```diff
@@ -79,20 +93,42 @@ public function validate( $post_id ) {
 
 		// Get active mappings.
 		$mappings = $this->dbo->get_active_mappings();
+		/**
+		 * Apply this filter to get mappings from external plugins.
+		 *
+		 * @param $mappings array Array of mappings from database.
+		 * @param $post_id int The post id.
+		 */
+		$mappings = apply_filters( 'wl_mappings_post', $mappings, $post_id );
 
 		// Get all active rule groups for the mapping items.
 		foreach ( $mappings as $mapping ) {
-			$rule_groups = $this->dbo->get_rule_groups_by_mapping( (int) $mapping['mapping_id'] );
-
-			$should_apply_mapping = $this->rule_groups_validator->is_valid( $post_id, $rule_groups );
-
-			if ( $should_apply_mapping ) {
-				$mapping_item_properties = $this->dbo->get_properties( $mapping['mapping_id'] );
-				$properties              = array_merge( $properties, $mapping_item_properties );
+			if ( array_key_exists( 'mapping_id', $mapping ) ) {
+				$rule_groups          = $this->dbo->get_rule_groups_by_mapping( (int) $mapping['mapping_id'] );
+				$should_apply_mapping = $this->rule_groups_validator->is_valid( $post_id, $rule_groups );
+				if ( $should_apply_mapping ) {
+					$mapping_item_properties = $this->dbo->get_properties( $mapping['mapping_id'] );
+					$properties              = array_merge( $properties, $mapping_item_properties );
+				}
+			} else {
+				/**
+				 * This is a programatically defined mapping,
+				 * so we will have the rule groups and the properties in the array keys
+				 */
+				if ( array_key_exists( 'rule_groups', $mapping )
+				     && array_key_exists( 'properties', $mapping ) &&
+				     is_array( $mapping['rule_groups'] ) &&
+				     is_array( $mapping['properties'] ) ) {
+					// Make sure we have the same data type.
+					$should_apply_mapping = $this->rule_groups_validator->is_valid( $post_id, $mapping['rule_groups'] );
```

Comment:
@naveen17797 you don't need this, the developer decides whether to apply the mappings by checking against the the post's ID provided in the filter; there are no rule groups, the developer provides only the mappings:

```apply_filters( 'wl_mappings_post', $mappings, $post_id );```